# Read file


In [2]:
import pandas as pd
from io import StringIO

with open("../data/CS_data.txt", "r", encoding="utf-8") as file:
    data = file.read()

# Load the CS_data.txt file into a DataFrame
data_df = pd.read_csv(StringIO(data), sep="\t")
print(type(data_df))

# Display the first few rows of the dataset to understand its structure
data_df.head()

<class 'pandas.core.frame.DataFrame'>


C:\Users\redhawc\AppData\Local\Temp\ipykernel_25324\2927799512.py:8: DtypeWarning: Columns (52,53) have mixed types. Specify dtype option on import or set low_memory=False.
  data_df = pd.read_csv(StringIO(data), sep="\t")


ID_CLIENT CLERK_TYPE  PAYMENT_DAY APPLICATION_SUBMISSION_TYPE  \
1          1          C       -99999                           0   
2          2          C       -99999                         Web   
3          3          C       -99999                           0   
4          4          C       -99999                         Web   
5          5          C       -99999                         Web   

   QUANT_ADDITIONAL_CARDS  POSTAL_ADDRESS_TYPE SEX  MARITAL_STATUS  \
1                       0                    1   F               6   
2                       0                    1   F               2   
3                       0                    1   F               2   
4                       0                    1   F               2   
5                       0                    1   M               2   

   QUANT_DEPENDANTS  EDUCATION_LEVEL  ... FLAG_HOME_ADDRESS_DOCUMENT FLAG_RG  \
1                 1              NaN  ...                          0       0   
2                 0              NaN  ...                          0       0   
3                 0              NaN  ...                          0       0   
4                 0              NaN  ...                          0       0   
5                 0              NaN  ...                          0       0   

   FLAG_CPF FLAG_INCOME_PROOF PRODUCT FLAG_ACSP_RECORD AGE  RESIDENCIAL_ZIP_3  \
1         0                 0       1                N  32                595   
2         0                 0       1                N  34                230   
3         0                 0       1                N  27                591   
4         0                 0       1                N  61                545   
5         0                 0       1                N  48                235   

   PROFESSIONAL_ZIP_3  TARGET_LABEL_BAD.1  
1                 595                   1  
2                 230                   1  
3                 591                   0  
4                 545                   0  
5                 235                   1  

[5 rows x 54 columns]

# KEEP ONLY THOSE USABLE VARIABLES


In [3]:
variables_list = [
    "PAYMENT_DAY",
    "RESIDENCIAL_STATE",
    "FLAG_RESIDENCIAL_PHONE",
    "MONTHS_IN_RESIDENCE",
    "PERSONAL_MONTHLY_INCOME",
    "OTHER_INCOMES",
    "AGE",
    "QUANT_CARS",
    "COMPANY",
    "FLAG_PROFESSIONAL_PHONE",
    "QUANT_DEPENDANTS",
    "APPLICATION_SUBMISSION_TYPE",
    "POSTAL_ADDRESS_TYPE",
    "NATIONALITY",
    "FLAG_VISA",
    "FLAG_MASTERCARD",
    "FLAG_DINERS",
    "FLAG_AMERICAN_EXPRESS",
    "MARITAL_STATUS",
]

print(len(variables_list))

data_df[variables_list].head()

19


PAYMENT_DAY RESIDENCIAL_STATE FLAG_RESIDENCIAL_PHONE  MONTHS_IN_RESIDENCE  \
1       -99999                RN                      Y                 15.0   
2       -99999                RJ                      Y                  1.0   
3       -99999                RN                      Y                  NaN   
4       -99999                PE                      N                  NaN   
5       -99999                RJ                      Y                 12.0   

   PERSONAL_MONTHLY_INCOME  OTHER_INCOMES  AGE  QUANT_CARS COMPANY  \
1                    900.0            0.0   32           0       N   
2                    750.0            0.0   34           0       Y   
3                    500.0            0.0   27           0       N   
4                    500.0            0.0   61           0       N   
5                   1200.0            0.0   48           0       N   

  FLAG_PROFESSIONAL_PHONE  QUANT_DEPENDANTS APPLICATION_SUBMISSION_TYPE  \
1                       N                 1                           0   
2                       N                 0                         Web   
3                       N                 0                           0   
4                       N                 0                         Web   
5                       N                 0                         Web   

   POSTAL_ADDRESS_TYPE  NATIONALITY  FLAG_VISA  FLAG_MASTERCARD  FLAG_DINERS  \
1                    1            1          1                1            0   
2                    1            1          0                0            0   
3                    1            1          0                0            0   
4                    1            1          0                0            0   
5                    1            1          0                0            0   

   FLAG_AMERICAN_EXPRESS  MARITAL_STATUS  
1                      0               6  
2                      0               2  
3                      0               2  
4                      0               2  
5                      0               2

In [4]:
data_df["PERSONAL_MONTHLY_INCOME"].value_counts()

PERSONAL_MONTHLY_INCOME
350.00     6898
500.00     1453
400.00     1402
380.00     1362
600.00     1198
           ... 
412.39        1
406.78        1
2432.00       1
969.49        1
719.38        1
Name: count, Length: 5495, dtype: int64

In [5]:
data_df = data_df[variables_list]

# data cleaning and pre-processing


In [6]:
# Display descriptive statistics of the dataset
data_descriptive_stats = data_df.describe(include="all")

# Display the descriptive statistics for the first few columns for brevity
data_descriptive_stats.iloc[:, :10]

PAYMENT_DAY RESIDENCIAL_STATE FLAG_RESIDENCIAL_PHONE  \
count   50000.000000             50000                  50000   
unique           NaN                27                      2   
top              NaN                SP                      Y   
freq             NaN              8773                  41809   
mean     -167.151640               NaN                    NaN   
std      4239.371262               NaN                    NaN   
min    -99999.000000               NaN                    NaN   
25%        10.000000               NaN                    NaN   
50%        10.000000               NaN                    NaN   
75%        15.000000               NaN                    NaN   
max        25.000000               NaN                    NaN   

        MONTHS_IN_RESIDENCE  PERSONAL_MONTHLY_INCOME  OTHER_INCOMES  \
count          46223.000000             50000.000000   50000.000000   
unique                  NaN                      NaN            NaN   
top                     NaN                      NaN            NaN   
freq                    NaN                      NaN            NaN   
mean               9.727149               886.678437      35.434760   
std               10.668841              7846.959327     891.515142   
min                0.000000                60.000000       0.000000   
25%                1.000000               360.000000       0.000000   
50%                6.000000               500.000000       0.000000   
75%               15.000000               800.000000       0.000000   
max              228.000000            959000.000000  194344.000000   

                AGE    QUANT_CARS COMPANY FLAG_PROFESSIONAL_PHONE  
count   50000.00000  50000.000000   50000                   50000  
unique          NaN           NaN       2                       2  
top             NaN           NaN       N                       N  
freq            NaN           NaN   27959                   36510  
mean       43.24852      0.336140     NaN                     NaN  
std        14.98905      0.472392     NaN                     NaN  
min         6.00000      0.000000     NaN                     NaN  
25%        31.00000      0.000000     NaN                     NaN  
50%        41.00000      0.000000     NaN                     NaN  
75%        53.00000      1.000000     NaN                     NaN  
max       106.00000      1.000000     NaN                     NaN

# handle missing values
identifying all the missing values according to variable list

In [20]:
# Re-reading the variable list from the provided Excel file
variable_list_df = pd.read_excel("../data/variable list.XLS.xls")

# Display the content of the variable list
variable_list_df[["Var_Title", "Var_Description", "Field_Content"]]

selected_variables = [
    "PAYMENT_DAY",
    "QUANT_DEPENDANTS",
    "CITY_OF_BIRTH",
    "RESIDENCIAL_STATE",
    "RESIDENCIAL_BOROUGH",
    "FLAG_RESIDENCIAL_PHONE",
    "MONTHS_IN_RESIDENCE",
    "FLAG_MOBILE_PHONE",
    "PERSONAL_MONTHLY_INCOME",
    "OTHER_INCOMES",
    "QUANT_BANKING_ACCOUNTS",
    "QUANT_SPECIAL_BANKING_ACCOUNTS",
    "PERSONAL_ASSETS_VALUE",
    "QUANT_CARS",
    "COMPANY",
    "FLAG_PROFESSIONAL_PHONE",
    "MONTHS_IN_THE_JOB",
    "FLAG_ACSP_RECORD",
    "AGE",
    "APPLICATION_SUBMISSION_TYPE",
    "POSTAL_ADDRESS_TYPE",
    "FLAG_EMAIL",
    "FLAG_VISA",
    "NATIONALITY",
    "FLAG_MASTERCARD",
    "FLAG_DINERS",
    "FLAG_AMERICAN_EXPRESS",
    "MARITAL_STATUS",
    "TARGET_LABEL_BAD.1",
]

filtered_df = variable_list_df.loc[
    variable_list_df["Var_Title"].isin(selected_variables)
]

filtered_df[["Var_Title", "Var_Description", "Field_Content"]]

Var_Title  \
2                      PAYMENT_DAY   
3      APPLICATION_SUBMISSION_TYPE   
5              POSTAL_ADDRESS_TYPE   
7                   MARITAL_STATUS   
8                 QUANT_DEPENDANTS   
11                   CITY_OF_BIRTH   
13               RESIDENCIAL_STATE   
15             RESIDENCIAL_BOROUGH   
16          FLAG_RESIDENCIAL_PHONE   
19             MONTHS_IN_RESIDENCE   
20               FLAG_MOBILE_PHONE   
21                      FLAG_EMAIL   
22         PERSONAL_MONTHLY_INCOME   
23                   OTHER_INCOMES   
24                       FLAG_VISA   
25                 FLAG_MASTERCARD   
26                     FLAG_DINERS   
27           FLAG_AMERICAN_EXPRESS   
29          QUANT_BANKING_ACCOUNTS   
30  QUANT_SPECIAL_BANKING_ACCOUNTS   
31           PERSONAL_ASSETS_VALUE   
32                      QUANT_CARS   
33                         COMPANY   
37         FLAG_PROFESSIONAL_PHONE   
39               MONTHS_IN_THE_JOB   
49                FLAG_ACSP_RECORD   
50                             AGE   

                                      Var_Description    Field_Content  
2   Day of the month for bill payment, chosen by t...  1,5,10,15,20,25  
3   Indicates if the application was submitted via...       Web, Carga  
5   Indicates if the address for posting is the ho...              1.2  
7                               Encoding not informed    1,2,3,4,5,6,7  
8                                                 NaN     0, 1, 2, ...  
11                                                NaN              NaN  
13                                 State of residence              NaN  
15                               Borough of residence              NaN  
16  Indicates if the applicant possesses a home phone              Y,N  
19            Time in the current residence in months   1,2,... , NULL  
20  Indicates if the applicant possesses a mobile ...              Y,N  
21  Indicates if the applicant possesses an e-mail...              0.1  
22  Applicant's personal regular monthly income in...              NaN  
23  Applicant's other incomes monthly averaged in ...              NaN  
24  Flag indicating if the applicant is a VISA cre...              0.1  
25  Flag indicating if the applicant is a MASTERCA...              0.1  
26  Flag indicating if the applicant is a SINERS c...              0.1  
27  Flag indicating if the applicant is an AMERICA...              0.1  
29                                                NaN            0,1,2  
30                                                NaN            0,1,2  
31  Total value of the personal possessions such a...              NaN  
32           Quantity of cars the applicant possesses              NaN  
33  If the applicant has supplied the name of the ...              Y,N  
37  Indicates if the professional phone number was...              Y,N  
39                  Time in the current job in months              NaN  
49  Flag indicating if the applicant has any previ...             Y, N  
50        Applicant's age at the moment of submission              NaN

In [8]:
# variable_list_df.to_json("variable_list_df.json")
# data_df.to_json("data_df.json")

In [9]:
data_df.isnull().sum()


PAYMENT_DAY                       0
RESIDENCIAL_STATE                 0
FLAG_RESIDENCIAL_PHONE            0
MONTHS_IN_RESIDENCE            3777
PERSONAL_MONTHLY_INCOME           0
OTHER_INCOMES                     0
AGE                               0
QUANT_CARS                        0
COMPANY                           0
FLAG_PROFESSIONAL_PHONE           0
QUANT_DEPENDANTS                  0
APPLICATION_SUBMISSION_TYPE       0
POSTAL_ADDRESS_TYPE               0
NATIONALITY                       0
FLAG_VISA                         0
FLAG_MASTERCARD                   0
FLAG_DINERS                       0
FLAG_AMERICAN_EXPRESS             0
MARITAL_STATUS                    0
dtype: int64

In [10]:
for i in data_df.columns:
    if len(data_df[i].value_counts().keys()) < 10:
        print("\n", data_df[i].value_counts().to_string())
        indices_for_value = variable_list_df["Var_Title"][
            variable_list_df["Var_Title"] == i
        ].index
        print(
            "\n",
            variable_list_df["Field_Content"][indices_for_value].to_string(),
            "\n----------------------",
        )


 PAYMENT_DAY
 10       20387
 15        8731
 5         7334
 25        7013
 20        4687
 1         1758
-99999       90

 2    1,5,10,15,20,25 
----------------------

 FLAG_RESIDENCIAL_PHONE
Y    41809
N     8191

 16    Y,N 
----------------------

 QUANT_CARS
0    33193
1    16807

 32    NaN 
----------------------

 COMPANY
N    27959
Y    22041

 33    Y,N 
----------------------

 FLAG_PROFESSIONAL_PHONE
N    36510
Y    13490

 37    Y,N 
----------------------

 APPLICATION_SUBMISSION_TYPE
Web      25232
0        19719
Carga     5049

 3    Web, Carga 
----------------------

 POSTAL_ADDRESS_TYPE
1    49673
2      327

 5    1.2 
----------------------

 NATIONALITY
1    47884
0     2018
2       98

 Series([], ) 
----------------------

 FLAG_VISA
0    44428
1     5572

 24    0.1 
----------------------

 FLAG_MASTERCARD
0    45127
1     4873

 25    0.1 
----------------------

 FLAG_DINERS
0    49934
1       66

 26    0.1 
----------------------

 FLAG_AMERICAN_EXPRE

        from above result, we can see that
# "MARITAL_STATUS" 
# "APPLICATION_SUBMISSION_TYPE"
# "PAYMENT_DAY"
# "NATIONALITY"
        contain hidden missing value

# "MONTHS_IN_RESIDENCE"
        contians obvious missing value

In [11]:
# Calculate the number of missing values for each column
missing_values = data_df.isnull().sum()

# Display columns with missing values and their counts
print(missing_values[missing_values > 0])

# Calculate the percentage of missing values for each column
missing_percentage = (data_df.isnull().sum() / len(data_df)) * 100
print(
    "----------------------------------------------\n",
    missing_percentage[missing_percentage > 50],
)

MONTHS_IN_RESIDENCE    3777
dtype: int64
----------------------------------------------
 Series([], dtype: float64)


# APPLICATION_SUBMISSION_TYPE
For the APPLICATION_SUBMISSION_TYPE variable, we see that there are three values: web, 0, and Carga. 

    The word "Carga" may be a Spanish or Portuguese word that can be translated as "load" or "charge".  

based on our variable descriptions, only web and Carga are valid, so we can reasonably assume that 0 represents a missing or incorrect value.

The strategy for dealing with this situation depends on our understanding of the data and the impact that the missing value may have on the model. Here are a few possible options:

Most Frequent Value Replacement: We can replace 0 with the most frequently occurring value. in this case, Web is the most common value, and we can replace all 0s with Web.

Categorical Replacement: If we assume that the 0s might represent some undocumented submission, we can categorise them as Carga, which is another known category in addition to Web.

`I used Most frequent value to replace NaN`



In [12]:
# Replace '0' values in 'APPLICATION_SUBMISSION_TYPE' with 'Web'
data_df["APPLICATION_SUBMISSION_TYPE"].replace("0", "Web", inplace=True)

data_df["APPLICATION_SUBMISSION_TYPE"].value_counts()

APPLICATION_SUBMISSION_TYPE
Web      44951
Carga     5049
Name: count, dtype: int64

# MONTHS_IN_RESIDENCE - replace by MEDIAN

The MONTHS_IN_RESIDENCE variable has unique values ranging from 0 to 228 months. Most of the values appear to be reasonable lengths of residence, but the maximum value of 228 months, or 19 years, while possible, may be noticeable. Usually, if this is not too common, we can keep these data as they may represent the reality of long-term residents.

With regard to dealing with missing values (NaN), we can use the `median` to fill in as the median is less sensitive to extreme values. In this dataset, the median is 6 months, which means that half of the applicants have lived in their current residence for less than or equal to 6 months. This is a reasonable value to represent a typical length of residence.

`I used median to replace NaN`


In [13]:
data_df["MONTHS_IN_RESIDENCE"].fillna(
    data_df["MONTHS_IN_RESIDENCE"].median(), inplace=True
)

data_df["MONTHS_IN_RESIDENCE"].isnull().sum()

0

# "MARITAL_STATUS" - replace by MODE

In the MARITAL_STATUS variable, we see that there are eight different values from 0 to 7. Based on the information you provided for the variable, we have marital statuses coded 1 through 7, but there is no mention of 0. Therefore, it is reasonable to assume that 0 represents a missing or incorrect value.

`Since we don't have detailed instructions to guide us on what to do with 0, I'm inclined to take the first strategy of replacing 0 with the most common marital status.This way, we can keep as much data as possible while still having a complete variable for marital status. Let's make the substitution. ​`

In [14]:
data_df["MARITAL_STATUS"].value_counts()

MARITAL_STATUS
2    25967
1    15286
4     4206
6     1877
5     1296
3      632
7      534
0      202
Name: count, dtype: int64

In [15]:
data_df['MARITAL_STATUS'].replace(0, data_df['MARITAL_STATUS'].mode()[0], inplace=True)

# Verify the replacement
data_df['MARITAL_STATUS'].value_counts()

MARITAL_STATUS
2    26169
1    15286
4     4206
6     1877
5     1296
3      632
7      534
Name: count, dtype: int64

# "PAYMENT_DAY"- replace by MODE

In the PAYMENT_DAY variable, there is a clear outlier of -99999, which is clearly not a valid payment day. This value occurs 90 times. We can assume that this is a placeholder for a missing or unknown payment day.

`Since we don't have additional information to determine which day might be the most appropriate replacement value, we can choose to use the most frequent payment day to replace the -99999 value. Let's find the most frequent payment day and then use it to replace the outlier.`

In [16]:
data_df["PAYMENT_DAY"][data_df["PAYMENT_DAY"] > 0].mode()[0]

# Replace the invalid payment day with the most frequent valid payment day
data_df["PAYMENT_DAY"].replace(
    -99999, data_df["PAYMENT_DAY"][data_df["PAYMENT_DAY"] > 0].mode()[0], inplace=True
)

# Verify the replacement
print(len(data_df["PAYMENT_DAY"]))
data_df["PAYMENT_DAY"].value_counts()

50000


PAYMENT_DAY
10    20477
15     8731
5      7334
25     7013
20     4687
1      1758
Name: count, dtype: int64

# "NATIONALITY" - PREFER TO DO NOTHING, WE CAN GO BACK HERE AFTER WE'VE BUILT THE MODEL


`Country of birth. Encoding not informed but Brazil is likely to be equal 1.` 

The NATIONALITY variable was distributed as follows:

Code 1 (probably representing Brazilian nationality) accounted for 95.768 per cent.
Code 0 accounted for 4.036 per cent.
Code 2 accounted for only 0.196 per cent.
Given that the vast majority of records are labelled as 1, we can presume that this is the dominant nationality category (most likely Brazilian). The small percentage of codes 0 and 2 suggests that they may represent other nationalities.

If we considered nationality to be an important feature in a credit scoring model, we might choose to retain these codes and allow the model to learn possible differences across these categories. However, as the sample size for code 2 is very small (0.196%), it may be reasonable to categorise it as 'other nationality' if there is no more information available, as this category may not have a significant impact on the model. The code 0, although a small percentage, may be a sufficient sample size to have an impact on the model and can therefore be retained.

`In summary, if nationality does matter for credit scoring, we can retain the status quo. If we determine that it is not a key feature, or if we want to reduce potential noise and complexity, we can categorise code 2 as 'other nationality' or even remove the variable altogether. `



In [17]:
data_df['NATIONALITY'].value_counts(normalize=True) * 100  # Get the distribution in percentage

NATIONALITY
1    95.768
0     4.036
2     0.196
Name: proportion, dtype: float64

In [18]:
grouped = data_df.groupby("NATIONALITY")
descriptive_stats = grouped.describe()

# 打印结果
descriptive_stats.to_csv("descriptive_stats.csv")

In [19]:
data_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 50000 entries, 1 to 50000
Data columns (total 19 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   PAYMENT_DAY                  50000 non-null  int64  
 1   RESIDENCIAL_STATE            50000 non-null  object 
 2   FLAG_RESIDENCIAL_PHONE       50000 non-null  object 
 3   MONTHS_IN_RESIDENCE          50000 non-null  float64
 4   PERSONAL_MONTHLY_INCOME      50000 non-null  float64
 5   OTHER_INCOMES                50000 non-null  float64
 6   AGE                          50000 non-null  int64  
 7   QUANT_CARS                   50000 non-null  int64  
 8   COMPANY                      50000 non-null  object 
 9   FLAG_PROFESSIONAL_PHONE      50000 non-null  object 
 10  QUANT_DEPENDANTS             50000 non-null  int64  
 11  APPLICATION_SUBMISSION_TYPE  50000 non-null  object 
 12  POSTAL_ADDRESS_TYPE          50000 non-null  int64  
 13  NATIONALITY          